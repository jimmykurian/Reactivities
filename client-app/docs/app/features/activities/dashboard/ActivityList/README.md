[**Client App Documentation v0.0.0**](../../../../../README.md) â€¢ **Docs**

***

[Client App Documentation v0.0.0](../../../../../README.md) / app/features/activities/dashboard/ActivityList

# app/features/activities/dashboard/ActivityList

## Index

### Functions

- [default](functions/default.md)
