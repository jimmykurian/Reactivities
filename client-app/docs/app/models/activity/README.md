[**Client App Documentation v0.0.0**](../../../README.md) â€¢ **Docs**

***

[Client App Documentation v0.0.0](../../../README.md) / app/models/activity

# app/models/activity

## Index

### Interfaces

- [Activity](interfaces/Activity.md)
