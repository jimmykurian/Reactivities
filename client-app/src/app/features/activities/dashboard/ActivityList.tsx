/**
 * @file ActivityList.tsx
 * @author Jimmy Kurian
 * @fileoverview This file contains the ActivityList component.
 */

import { Button, Item, Label, Segment } from 'semantic-ui-react';
import { SyntheticEvent, useState } from 'react';
import { useStore } from '../../../stores/store';
import { observer } from 'mobx-react-lite';

/**
 * @component ActivityList
 * @description The ActivityList component displays a list of activities with options to view and delete each activity.
 *
 * @returns {JSX.Element} The rendered ActivityList component.
 *
 * @remarks
 * This component maps over an array of activities and displays each one with its details using Semantic UI components.
 * Each activity includes a title, date, description, city, venue, and category, along with "View" and "Delete" buttons.
 * The `activityStore.selectActivity` function is called when the "View" button is clicked, passing the activity's ID.
 * The `deleteActivity` function is called when the "Delete" button is clicked, passing the activity's ID.
 * The `loading` state is used to indicate whether a delete operation is in progress, which disables the delete button for the targeted activity.
 *
 * @example
 * ```tsx
 * const activities = [
 *   { id: '1', title: 'Activity 1', date: '2023-12-31', description: 'Description 1', category: 'Category 1', city: 'City 1', venue: 'Venue 1' },
 *   { id: '2', title: 'Activity 2', date: '2024-01-01', description: 'Description 2', category: 'Category 2', city: 'City 2', venue: 'Venue 2' },
 * ];
 *
 * const deleteActivity = (id: string) => console.log(`Delete activity with id ${id}`);
 *
 * <ActivityList />
 * ```
 */
export default observer(function ActivityList(): JSX.Element {
  const { activityStore } = useStore();
  const { activities, deleteActivity, loading } = activityStore;
  const [target, setTarget] = useState('');

  /**
   * @function handleActivityDelete
   * @description Handles the deletion of an activity.
   * @param {SyntheticEvent<HTMLButtonElement>} e - The synthetic event triggered by clicking the delete button.
   * @param {string} id - The ID of the activity to delete.
   */
  function handleActivityDelete(
    e: SyntheticEvent<HTMLButtonElement>,
    id: string,
  ) {
    setTarget(e.currentTarget.name);
    deleteActivity(id);
  }

  return (
    <Segment>
      <Item.Group divided>
        {activities.map((activity) => (
          <Item key={activity.id}>
            <Item.Content>
              <Item.Header as="a">{activity.title}</Item.Header>
              <Item.Meta>{activity.date}</Item.Meta>
              <Item.Description>
                <div>{activity.description}</div>
                <div>
                  {activity.city}, {activity.venue}
                </div>
              </Item.Description>
              <Item.Extra>
                <Button
                  onClick={() => activityStore.selectActivity(activity.id)}
                  floated="right"
                  content="View"
                  color="blue"
                />
                <Button
                  name={activity.id}
                  loading={loading && target === activity.id}
                  onClick={(e) => handleActivityDelete(e, activity.id)}
                  floated="right"
                  content="Delete"
                  color="red"
                />
                <Label basic content={activity.category} />
              </Item.Extra>
            </Item.Content>
          </Item>
        ))}
      </Item.Group>
    </Segment>
  );
});
