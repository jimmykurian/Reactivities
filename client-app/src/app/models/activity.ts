/**
 * @author Jimmy Kurian
 * @name Activity
 */

/**
 * The Activity interface represents the structure of an activity object in the Reactivities application.
 * It includes various properties such as id, title, date, description, category, city, and venue.
 *
 * @interface
 *
 * @property {string} id - The unique identifier for the activity.
 * @property {string} title - The title of the activity.
 * @property {string} date - The date of the activity in ISO string format.
 * @property {string} description - A brief description of the activity.
 * @property {string} category - The category to which the activity belongs.
 * @property {string} city - The city where the activity takes place.
 * @property {string} venue - The specific venue of the activity within the city.
 */
export interface Activity {
  id: string;
  title: string;
  date: string;
  description: string;
  category: string;
  city: string;
  venue: string;
}
