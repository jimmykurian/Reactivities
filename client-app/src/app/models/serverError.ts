/**
 * @file ServerError.ts
 * @author Jimmy Kurian
 * @fileoverview This file defines the `ServerError` interface, which models the structure of server error responses in the Reactivities application.
 */

/**
 * The ServerError interface represents the structure of an error returned from the server.
 * It includes the HTTP status code, an error message, and optional detailed information about the error.
 *
 * @interface ServerError
 *
 * @property {number} statusCode - The HTTP status code associated with the error.
 * @property {string} message - A brief message describing the error.
 * @property {string} details - Additional details or stack trace information about the error (optional).
 */
export interface ServerError {
  statusCode: number;
  message: string;
  details: string;
}
