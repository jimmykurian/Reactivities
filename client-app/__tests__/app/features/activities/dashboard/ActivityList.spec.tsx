import { faker } from '@faker-js/faker';
import ActivityList from '../../../../../src/app/features/activities/dashboard/ActivityList';
import { Activity } from '../../../../../src/app/models/activity';
import { render, screen } from '@testing-library/react';

describe('ActivityList', () => {
  // Generate mock activities using Faker.js
  const generateMockActivity = (): Activity => ({
    id: faker.string.uuid(),
    title: faker.lorem.words(3),
    date: faker.date.future().toISOString(),
    description: faker.lorem.sentence(),
    category: faker.lorem.word(),
    city: faker.location.city(),
    venue: faker.location.streetAddress(),
  });

  // Generate mock activities dynamically for other tests
  const mockActivities: Activity[] = Array.from(
    { length: 2 },
    generateMockActivity,
  );

  // Fixed mock activities for snapshot test
  const fixedMockActivities: Activity[] = [
    {
      id: '1',
      title: 'Fixed Title 1',
      date: '2023-12-31T23:59:59.000Z',
      description: 'Fixed description 1',
      category: 'Fixed category 1',
      city: 'Fixed city 1',
      venue: 'Fixed venue 1',
    },
    {
      id: '2',
      title: 'Fixed Title 2',
      date: '2024-01-01T23:59:59.000Z',
      description: 'Fixed description 2',
      category: 'Fixed category 2',
      city: 'Fixed city 2',
      venue: 'Fixed venue 2',
    },
  ];

  const selectActivity = jest.fn();
  const deleteActivity = jest.fn();

  test('renders the ActivityList component', () => {
    // Arrange
    render(
      <ActivityList
        activities={mockActivities}
        selectActivity={selectActivity}
        deleteActivity={deleteActivity}
      />,
    );

    // Act & Assert
    mockActivities.forEach((activity) => {
      const activityElement = screen.getByText(activity.title);
      expect(activityElement).toBeInTheDocument();
    });
  });

  test('matches snapshot', () => {
    // Arrange
    const { asFragment } = render(
      <ActivityList
        activities={fixedMockActivities}
        selectActivity={selectActivity}
        deleteActivity={deleteActivity}
      />,
    );

    // Act & Assert
    expect(asFragment()).toMatchSnapshot();
  });
});
