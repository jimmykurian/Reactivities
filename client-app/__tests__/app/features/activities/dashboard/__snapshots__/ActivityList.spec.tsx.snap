// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityList matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui teal sub header"
  >
    2023-12-31
  </div>
  <div
    class="ui segments"
  >
    <div
      class="ui segment"
    >
      <div
        class="ui items"
      >
        <div
          class="item"
        >
          <div
            class="ui tiny circular image"
          >
            <img
              src="/assets/user.png"
            />
          </div>
          <div
            class="content"
          >
            <a
              class="header"
              href="/activities/1"
            >
              Fixed Title 1
            </a>
            <div
              class="description"
            >
              Hosted by Bob
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ui segment"
    >
      <span>
        <i
          aria-hidden="true"
          class="clock icon"
        />
         2023-12-31
        <i
          aria-hidden="true"
          class="marker icon"
        />
         Fixed venue 1
      </span>
    </div>
    <div
      class="ui secondary segment"
    >
      Attendees will go here
    </div>
    <div
      class="ui clearing segment"
    >
      <span>
        Fixed description 1
      </span>
      <a
        class="ui teal right floated button"
        href="/activities/1"
        role="button"
      >
        View
      </a>
      <button
        class="ui red right floated button"
        name="1"
      >
        Delete
      </button>
    </div>
  </div>
  <div
    class="ui teal sub header"
  >
    2024-01-01
  </div>
  <div
    class="ui segments"
  >
    <div
      class="ui segment"
    >
      <div
        class="ui items"
      >
        <div
          class="item"
        >
          <div
            class="ui tiny circular image"
          >
            <img
              src="/assets/user.png"
            />
          </div>
          <div
            class="content"
          >
            <a
              class="header"
              href="/activities/2"
            >
              Fixed Title 2
            </a>
            <div
              class="description"
            >
              Hosted by Bob
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ui segment"
    >
      <span>
        <i
          aria-hidden="true"
          class="clock icon"
        />
         2024-01-01
        <i
          aria-hidden="true"
          class="marker icon"
        />
         Fixed venue 2
      </span>
    </div>
    <div
      class="ui secondary segment"
    >
      Attendees will go here
    </div>
    <div
      class="ui clearing segment"
    >
      <span>
        Fixed description 2
      </span>
      <a
        class="ui teal right floated button"
        href="/activities/2"
        role="button"
      >
        View
      </a>
      <button
        class="ui red right floated button"
        name="2"
      >
        Delete
      </button>
    </div>
  </div>
</DocumentFragment>
`;
